### PR TITLE
adding the 1.data-exploration module as well as the file sample_info_…

### PR DESCRIPTION
**Pull request by mjw-18**
…daya_visualization.ipynb to visualize the sample_info.csv data file. the environment.yml file has also been updated to reflect the new packages used in the new code

### CODE REVIEW

**Comment by mjw-18** [comment]
@gwaybio hey greg ive submitted a few changes from my branch named prime.

File: 1.data-exploration/sample_info_data_visualization.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+# assign the desired file a variable using pathlib.Path command
+input_file = pathlib.Path("../0.data-download/data/sample_info.csv")
+
+# set the data frame to be the desired .csv file that is read by pandas(pd) using the pd.read_csv(desired file read as a previously defined variable)
+df = pd.read_csv(input_file)
+
+# print the parameters of the read file
+print(df.shape)
+
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: it's also good practice to always add a `df.shape(3)` (or something other than 3) here to show the structure of the input data.

File: 1.data-exploration/sample_info_data_visualization.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+# In[2]:
+
+
+# how many samples?
+n_samples = len(df["DepMap_ID"])
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: is `DepMap_ID` always going to be a unique value?

perhaps something like this is better:

```python
n_samples = len(df["DepMap_ID"].unique())
```

File: 1.data-exploration/sample_info_data_visualization.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+print(f"Number of Samples: {n_samples} \n")
+
+# how many genes
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: The comment says genes, but I think you mean different ages?

File: 1.data-exploration/sample_info_data_visualization.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+# In[2]:
+
+
+# how many samples?
+n_samples = len(df["DepMap_ID"])
+print(f"Number of Samples: {n_samples} \n")
+
+# how many genes 
+all_cancers = df["age"].unique()
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: this variable should be `all_ages`, correct?

And need to fix in the next line too :)

File: 1.data-exploration/sample_info_data_visualization.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+all_cancers = df["primary_disease"].unique()
+print(f"All Cancer Types: \n {all_cancers} \n")
+
+
+# In[3]:
+
+
+# "Adult" was set to 18
+# "Pediatric" was set to 0
+# "Fetus" was set to 0
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: Is this a common metadata distinction?

I have never seen "Fetus" before, so this is a bit surprising to me

File: 1.data-exploration/sample_info_data_visualization.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+no_nan_age_df = df.loc[df["age"].notnull()]
+
+# create a new data frame that will also assign the integer value of 18 to cells containing "Adult"
+no_nan_age_df.loc[(df["age"] == "Adult")] = 18
+
+# create a new data frame that will also assign the integer value of 0 to cells containing "Pediatric"
+no_nan_age_df.loc[(df["age"] == "Pediatric")] = 0
+
+# create a new data frame that will also assign the integer value of 0 to cells containg "Fetus"
+no_nan_age_df.loc[(df["age"] == "Fetus")] = -1
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: Oh, I see. You should look around on the DepMap download page for some sort of README documenting details of the file.

-1 usually means _missing data_ in this context.

File: 1.data-exploration/sample_info_data_visualization.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+# In[4]:
+
+
+ig, axs = plt.subplots(1, 2, dpi=150)
+
+# plot adult age distribution 
+sns.histplot(x="age", data=adult_df, stat="frequency", ax=axs[0])
+
+# plot pediatric age distribution
+sns.histplot(x="age", data=pediatric_df, stat="count", ax=axs[1])
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: I like how you're thinking, but it's best practice to not split out data into two figures if you can help it, especially if splitting creates a different y axis.

Instead, can you create one plot and then add a line of x=18? indicating adult vs. pediatric.

I think something like this would work:

```python
import plotnine as gg

data = ....LOAD DATA HERE....

adult_age = 18

(
    gg.ggplot(data, gg.aes(x="age"))
    + gg.geom_histogram()
    + gg.geom_vline(xintercept = adult_age, color = "red", linetype = "dashed")
)
```

File: 1.data-exploration/sample_info_data_visualization.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+# create a new data frame that will also assign the integer value of 0 to cells containg "Fetus"
+no_nan_age_df.loc[(df["age"] == "Fetus")] = -1
+
+
+
+adult_df = no_nan_age_df.loc[no_nan_age_df["age"].astype(int) >= 18]
+adult_df = adult_df.astype({"age": int})
+pediatric_df = no_nan_age_df.loc[(no_nan_age_df["age"].astype(int) < 18) & (no_nan_age_df["age"].astype(int) != -1)]
+pediatric_df = pediatric_df.astype({"age": int})
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: See comment below, I don't think you need to do this

File: 1.data-exploration/sample_info_data_visualization.py (hunk 5 above)
Comment (gwaybio) [reply 1.data-exploration/sample_info_data_visualization.py]: Also, having these comments is great, but you can also code this explicitly, and then use a variable _instead of hardcoding_ later on in the script. This is more pythonic to do it this way :)

for example:

```python
adult_age = 18
ped_age = 0

# create a new data frame that will also assign the integer value of 18 to cells containing "Adult"
no_nan_age_df.loc[(df["age"] == "Adult")] = adult_age

.... and so on...
```

File: 1.data-exploration/sample_info_data_visualization.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+n_samples = len(df["DepMap_ID"])
+print(f"Number of Samples: {n_samples} \n")
+
+# how many genes 
+all_cancers = df["age"].unique()
+print(f"Ages sampled from: \n {all_cancers} \n")
+
+# how many different types of cancer?
+all_cancers = df["primary_disease"].unique()
+print(f"All Cancer Types: \n {all_cancers} \n")
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: Knowing which cancer types are present is a great first step, but how many of each are there?

A bar chart for this would be great :)

Something like:

```
import plotnine as gg

data = ....LOAD DATA HERE....

(
    gg.ggplot(data, gg.aes(x="primary_disease"))
    + gg.geom_barchart()
)
```

You may also need to code a new variable called "other" in case there are many singleton cancer types.

File: environment.yml [hunk 10]
```diff
@@ -5,3 +5,5 @@ dependencies:
 - conda-forge::jupyterlab	=	3.4.3
 - conda-forge::python	=	3.9.13
 - conda-forge::urllib3	=	1.26.9
+- conda-forge::seaborn = 0.11.2
+- conda-forge::plotnine = 0.8.0
```
Comment (gwaybio) [inline environment.yml]:
```suggestion
- conda-forge::plotnine = 0.8.0

```

Remember to add new line. I wonder if `vscode` doesn't do this automatically 🤔

File: 1.data-exploration/sample_info_data_visualization.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+# visualizes the general distribution of ages sampled from
+
+
+# In[5]:
+
+
+pediatric_df.loc[pediatric_df["age"] == 1]
+
+
+# In[ ]:
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: Make sure to delete empty cells (this is a different policy than having an extra new line at end of other files!) :)

File: 1.data-exploration/sample_info_data_visualization.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python
```
Comment (gwaybio) [inline 1.data-exploration/sample_info_data_visualization.py]: please make sure when you `nbconvert` you send it to a folder called `scripts/nbconverted` like you did in 0.data-download.

```bash
jupyter nbconvert --to=script --FilesWriter.build_directory=scripts/nbconverted *.ipynb
```

File: 0.data-download/data_downloader.ipynb [hunk 13]
```diff
@@ -92,6 +92,13 @@
     "        output_file=output_file\n",
     "    )"
    ]
+  },
```
Comment (gwaybio) [inline 0.data-download/data_downloader.ipynb]: hmm, looks like you added this by mistake. Can you clean/remove?

**Review by gwaybio — CHANGES_REQUESTED** [review]
Looking great! Several discussion items below.

One major item I didn't see was a confirmation that all `DepMap_ID`s aligned with the `gene_dependency` data. Sometimes all metadata samples are not actually measured, so it will be good to confirm that the ages and cancer types you're reporting are actually measured.

TLDR you'll need to cross reference IDs between files to make sure all samples with metadata are actually used.

File: 1.data-exploration/sample_info_data_visualization.py (hunk 1 above)
Comment (mjw-18) [reply 1.data-exploration/sample_info_data_visualization.py]: when i try and add a number inside of this i get an error. The output of df.shape describes how many rows and columns are within the data set

File: 1.data-exploration/sample_info_data_visualization.py (hunk 4 above)
Comment (mjw-18) [reply 1.data-exploration/sample_info_data_visualization.py]: this is correct i forgot to rename when i copy pasted the lines

File: 1.data-exploration/sample_info_data_visualization.py (hunk 5 above)
Comment (mjw-18) [reply 1.data-exploration/sample_info_data_visualization.py]: I have not seen fetus before either. the assignment of the integer value of -1 to the string value of "Fetus" as it appears in the data is to create a unique integer value that has no effect on the real data and can therefor be ignored in the later function.

File: 1.data-exploration/sample_info_data_visualization.py (hunk 7 above)
Comment (mjw-18) [reply 1.data-exploration/sample_info_data_visualization.py]: this was actually a mistake i did. the Y axis is the same for both graphs and should be "count". I added "frequency" in there as an experiment to test histplot parameters and forgot to change it back.

File: environment.yml (hunk 10 above)
Comment (mjw-18) [reply environment.yml]: i did forget to add another line! vscode evidently does not add this new line and so ill remember this in the future!

File: 1.data-exploration/sample_info_data_visualization.py (hunk 12 above)
Comment (mjw-18) [reply 1.data-exploration/sample_info_data_visualization.py]: should there be a scripts directory in each module for each modules respective codes?

File: 1.data-exploration/sample_info_data_visualization.py (hunk 1 above)
Comment (gwaybio) [reply 1.data-exploration/sample_info_data_visualization.py]: ah! Sorry, I meant `df.head(3)` 😊

File: 1.data-exploration/sample_info_data_visualization.py (hunk 5 above)
Comment (gwaybio) [reply 1.data-exploration/sample_info_data_visualization.py]: gotcha! If it can be ignored, then I think we can set it to "ignored" instead of "Fetus".

Also (and I think I mentioned this later on), it'll be good to find some sort of data manifest (might be in a README or elsewhere on the DepMap download) that describes these codes. For example, -1 might actually mean missing.

If you can find where these data are codified, please add to this pull request. Having accurate metadata labels is _extremely_ important! :)

File: 1.data-exploration/sample_info_data_visualization.py (hunk 7 above)
Comment (gwaybio) [reply 1.data-exploration/sample_info_data_visualization.py]: It is still a bit strange to me seeing two panels when it could easily be described simply in one

File: 1.data-exploration/sample_info_data_visualization.py (hunk 12 above)
Comment (gwaybio) [reply 1.data-exploration/sample_info_data_visualization.py]: yes, but not always. 

`scripts/` stores any additional custom functions that you write, and a `scripts/nbconverted` folder that stores all notebooks you convert to `.py`.

For example, when you start coding the B-VAE, you will write a `2.train-vae/scripts/vae.py` script that stores all custom functions that you then import in `2.train-vae/vae_training.ipynb`. (you will also have a `2.train-vae/scripts/nbconverted/vae_training.py`)

**Comment by mjw-18** [comment]
@gwaybio hey greg this pull request is ready for review!

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+
+# In[2]:
+
+
+# assign the desired file a variable using pathlib.Path command
+input_file = pathlib.Path("../0.data-download/data/sample_info.csv")
+
+# set the data frame to be the desired .csv file that is read by pandas(pd) using the pd.read_csv(desired file read as a previously defined variable)
+df = pd.read_csv(input_file)
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]: I recommend naming conventions that better indicate what is stored in the variable.

```suggestion
sample_info_df = pd.read_csv(input_file)
```

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+# assign the desired file a variable using pathlib.Path command
+input_file2 = pathlib.Path("../0.data-download/data/CRISPR_gene_dependency.csv")
+
+# set the data frame to be the desired .csv file that is read by pandas(pd) using the pd.read_csv(desired file read as a previously defined variable)
+df2 = pd.read_csv(input_file2)
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]: Also here:

```suggestion
dependency_df = pd.read_csv(input_file2)
```

(or `x_df`)

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+
+# In[3]:
+
+
+# print the parameters of the read file
+print(df.shape)
+df.head(1)
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]: This is beautiful to see 😄

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+# In[5]:
+
+
+# sample_info.csv visualization
+# how many samples from sample_info.csv?
+n_samples = len(df["DepMap_ID"].unique())
+print(f"Number of Samples Documented in sample_info.csv: {n_samples} \n")
+
+# sample_info.csv visualization
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]:
```suggestion
```

I think this was added in error

File: environment.yml [hunk 18]
```diff
@@ -5,3 +5,6 @@ dependencies:
...
+- conda-forge::tensorflow = 2.9.1
```
Comment (gwaybio) [inline environment.yml]: Does this PR use tensorflow? I didn't think so

**Review by gwaybio — CHANGES_REQUESTED** [review]
Really only minor discussion items. Please do not "accept" suggestions, as I made them in the `.py` file (that was built using `nbconvert`).

Instead, make the appropriate changes in the corresponding `.ipynb` file and then reconvert.

Also, please make sure to save your figures! Check out [`plotnine.save`](https://plotnine.readthedocs.io/en/stable/generated/plotnine.ggplot.html#plotnine.ggplot.save).

(save your figures in a new folder called `1.data-exploration/figures`)

Nice job! I'm seeing lots of growth + maturity here!

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+
+# In[9]:
+
+
+age_categories_bar = (
+    gg.ggplot(df_age_visual, gg.aes(x="age_categories"))
+    + gg.geom_bar()
+)
+print(age_categories_bar)
+age_categories_bar.save("./figures/age_categories_bar_chart.png")
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]:
```suggestion
bar_output_file = pathlib.Path("./figures/age_categories_bar_chart.png")
age_categories_bar.save(bar_output_file)
```

Make sure all paths are predefined using pathlib.

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+# In[10]:
+
+
+age_distribution_plot = (
+    gg.ggplot(df_age_visual, gg.aes(x="age_distribution"))
+    + gg.geom_density()
+    + gg.geom_vline(xintercept=adult_threshold, linetype="dashed", color="red")
+)
+print(age_distribution_plot)
+age_distribution_plot.save("./figures/sample_age_distribution_plot.png")
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]: same comment as above

File: 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+
+# In[12]:
+
+
+gendersamp = df_sample_info
+gendersamp_plot = (
+    gg.ggplot(gendersamp, gg.aes(x="sex")) + gg.geom_bar()
+)
+print(gendersamp_plot)
+gendersamp_plot.save("./figures/sample_gender_bar_chart.png")
```
Comment (gwaybio) [inline 1.data-exploration/scripts/nbconverted/sample_info_data_visualization.py]: same comment as above.

**Review by gwaybio — APPROVED** [review]
One very minor comment.

After you address, feel free to merge.
